File: VB/App_Code/DataHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Data.OleDb
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports DevExpress.XtraScheduler
Imports DevExpress.Web.ASPxScheduler
Imports DevExpress.Web.ASPxScheduler.Internal

''' <summary>
''' Summary description for DataHelper
''' </summary>
Public NotInheritable Class DataHelper
	Private Sub New()
	End Sub
	Public Shared Sub SetupMappings(ByVal control As ASPxScheduler)
		Dim storage As ASPxSchedulerStorage = control.Storage
		storage.BeginUpdate()
		Try
			Dim resourceMappings As ASPxResourceMappingInfo = storage.Resources.Mappings
			resourceMappings.ResourceId = "Id"
			resourceMappings.Caption = "Caption"

			Dim appointmentMappings As ASPxAppointmentMappingInfo = storage.Appointments.Mappings
			appointmentMappings.AppointmentId = "Id"
			appointmentMappings.Start = "StartTime"
			appointmentMappings.End = "EndTime"
			appointmentMappings.Subject = "Subject"
			appointmentMappings.AllDay = "AllDay"
			appointmentMappings.Description = "Description"
			appointmentMappings.Label = "Label"
			appointmentMappings.Location = "Location"
			appointmentMappings.RecurrenceInfo = "RecurrenceInfo"
			appointmentMappings.ReminderInfo = "ReminderInfo"
			appointmentMappings.ResourceId = "OwnerId"
			appointmentMappings.Status = "Status"
			appointmentMappings.Type = "EventType"
		Finally
			storage.EndUpdate()
		End Try
	End Sub
	Public Shared Sub ProvideRowInsertion(ByVal control As ASPxScheduler, ByVal dataSource As DataSourceControl)
		Dim accessDataSource As AccessDataSource = TryCast(dataSource, AccessDataSource)
		If accessDataSource IsNot Nothing Then
			Dim provider As New AccessRowInsertionProvider()
			provider.ProvideRowInsertion(control, accessDataSource)
			Return
		End If
		Dim objectDataSource As ObjectDataSource = TryCast(dataSource, ObjectDataSource)
		If objectDataSource IsNot Nothing Then
			Dim provider As New ObjectDataSourceRowInsertionProvider()
			provider.ProvideRowInsertion(control, objectDataSource)
		End If
	End Sub
End Class
Public Class AccessRowInsertionProvider
	Private lastInsertedAppointmentId As Integer

	Public Sub ProvideRowInsertion(ByVal control As ASPxScheduler, ByVal dataSource As AccessDataSource)
		AddHandler dataSource.Inserted, AddressOf AppointmentsDataSource_Inserted
		AddHandler control.AppointmentRowInserting, AddressOf ControlOnAppointmentRowInserting
		AddHandler control.AppointmentRowInserted, AddressOf ControlOnAppointmentRowInserted
		AddHandler control.AppointmentsInserted, AddressOf ControlOnAppointmentsInserted
	End Sub

	Private Sub ControlOnAppointmentRowInserting(ByVal sender As Object, ByVal e As ASPxSchedulerDataInsertingEventArgs)
		' Autoincremented primary key case
		e.NewValues.Remove("ID")
	End Sub
	Private Sub ControlOnAppointmentRowInserted(ByVal sender As Object, ByVal e As ASPxSchedulerDataInsertedEventArgs)
		' Autoincremented primary key case
		e.KeyFieldValue = Me.lastInsertedAppointmentId
	End Sub
	Private Sub AppointmentsDataSource_Inserted(ByVal sender As Object, ByVal e As SqlDataSourceStatusEventArgs)
		' Autoincremented primary key case
		Dim connection As OleDbConnection = CType(e.Command.Connection, OleDbConnection)
		Using cmd As New OleDbCommand("SELECT @@IDENTITY", connection)
			Me.lastInsertedAppointmentId = CInt(Fix(cmd.ExecuteScalar()))
		End Using
	End Sub
	Private Sub ControlOnAppointmentsInserted(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
		' Autoincremented primary key case
		Dim count As Integer = e.Objects.Count
		System.Diagnostics.Debug.Assert(count = 1)
		Dim apt As Appointment = CType(e.Objects(0), Appointment)
		Dim storage As ASPxSchedulerStorage = CType(sender, ASPxSchedulerStorage)
		storage.SetAppointmentId(apt, lastInsertedAppointmentId)
	End Sub
End Class
Public Class ObjectDataSourceRowInsertionProvider

	Public Sub ProvideRowInsertion(ByVal control As ASPxScheduler, ByVal dataSource As ObjectDataSource)
		AddHandler control.AppointmentInserting, AddressOf ControlOnAppointmentInserting
	End Sub
	Protected Sub ControlOnAppointmentInserting(ByVal sender As Object, ByVal e As PersistentObjectCancelEventArgs)
		Dim storage As ASPxSchedulerStorage = CType(sender, ASPxSchedulerStorage)
		Dim apt As Appointment = CType(e.Object, Appointment)
		storage.SetAppointmentId(apt, "a" & apt.GetHashCode())
	End Sub
End Class

File: CS/App_Code/DataHelper.cs
using System;
using System.Data.OleDb;
using System.Web.UI;
using System.Web.UI.WebControls;
using DevExpress.XtraScheduler;
using DevExpress.Web.ASPxScheduler;
using DevExpress.Web.ASPxScheduler.Internal;

/// <summary>
/// Summary description for DataHelper
/// </summary>
public static class DataHelper {
	public static void SetupMappings(ASPxScheduler control) {
		ASPxSchedulerStorage storage = control.Storage;
		storage.BeginUpdate();
		try {
			ASPxResourceMappingInfo resourceMappings = storage.Resources.Mappings;
			resourceMappings.ResourceId = "Id";
			resourceMappings.Caption = "Caption";

			ASPxAppointmentMappingInfo appointmentMappings = storage.Appointments.Mappings;
			appointmentMappings.AppointmentId = "Id";
			appointmentMappings.Start = "StartTime";
			appointmentMappings.End = "EndTime";
			appointmentMappings.Subject = "Subject";
			appointmentMappings.AllDay = "AllDay";
			appointmentMappings.Description = "Description";
			appointmentMappings.Label = "Label";
			appointmentMappings.Location = "Location";
			appointmentMappings.RecurrenceInfo = "RecurrenceInfo";
			appointmentMappings.ReminderInfo = "ReminderInfo";
			appointmentMappings.ResourceId = "OwnerId";
			appointmentMappings.Status = "Status";
			appointmentMappings.Type = "EventType";
		}
		finally {
			storage.EndUpdate();
		}
	}
	public static void ProvideRowInsertion(ASPxScheduler control, DataSourceControl dataSource) {
		AccessDataSource accessDataSource = dataSource as AccessDataSource;
		if(accessDataSource != null) {
			AccessRowInsertionProvider provider = new AccessRowInsertionProvider();
			provider.ProvideRowInsertion(control, accessDataSource);
			return;
		}
		ObjectDataSource objectDataSource = dataSource as ObjectDataSource;
		if(objectDataSource != null) {
			ObjectDataSourceRowInsertionProvider provider = new ObjectDataSourceRowInsertionProvider();
			provider.ProvideRowInsertion(control, objectDataSource);
		}
	}
}
public class AccessRowInsertionProvider {
	int lastInsertedAppointmentId;

	public void ProvideRowInsertion(ASPxScheduler control, AccessDataSource dataSource) {
		dataSource.Inserted += new SqlDataSourceStatusEventHandler(AppointmentsDataSource_Inserted);
		control.AppointmentRowInserting += new ASPxSchedulerDataInsertingEventHandler(ControlOnAppointmentRowInserting);
		control.AppointmentRowInserted += new ASPxSchedulerDataInsertedEventHandler(ControlOnAppointmentRowInserted);
		control.AppointmentsInserted += new PersistentObjectsEventHandler(ControlOnAppointmentsInserted);
	}

	void ControlOnAppointmentRowInserting(object sender, ASPxSchedulerDataInsertingEventArgs e) {
		// Autoincremented primary key case
		e.NewValues.Remove("ID");
	}
	void ControlOnAppointmentRowInserted(object sender, ASPxSchedulerDataInsertedEventArgs e) {
		// Autoincremented primary key case
		e.KeyFieldValue = this.lastInsertedAppointmentId;
	}
	void AppointmentsDataSource_Inserted(object sender, SqlDataSourceStatusEventArgs e) {
		// Autoincremented primary key case
		OleDbConnection connection = (OleDbConnection)e.Command.Connection;
		using(OleDbCommand cmd = new OleDbCommand("SELECT @@IDENTITY", connection)) {
			this.lastInsertedAppointmentId = (int)cmd.ExecuteScalar();
		}
	}
	void ControlOnAppointmentsInserted(object sender, PersistentObjectsEventArgs e) {
		// Autoincremented primary key case
		int count = e.Objects.Count;
		System.Diagnostics.Debug.Assert(count == 1);
		Appointment apt = (Appointment)e.Objects[0];
		ASPxSchedulerStorage storage = (ASPxSchedulerStorage)sender;
		storage.SetAppointmentId(apt, lastInsertedAppointmentId);
	}
}
public class ObjectDataSourceRowInsertionProvider {

	public void ProvideRowInsertion(ASPxScheduler control, ObjectDataSource dataSource) {
		control.AppointmentInserting += new PersistentObjectCancelEventHandler(ControlOnAppointmentInserting);
	}
	protected void ControlOnAppointmentInserting(object sender, PersistentObjectCancelEventArgs e) {
		ASPxSchedulerStorage storage = (ASPxSchedulerStorage)sender;
		Appointment apt = (Appointment)e.Object;
		storage.SetAppointmentId(apt, "a" + apt.GetHashCode());
	}
}

